include/wlr/types/wlr_buffer.h

```diff
@@ -21,23 +21,51 @@ struct wlr_buffer_impl {
 		struct wlr_dmabuf_attributes *attribs);
 };
 
+/**
+ * A buffer containing pixel data.
+ *
+ * A buffer has a single producer (the party who created the buffer) and
+ * multiple consumers (parties reading the buffer). When all consumers are done
+ * with the buffer, it gets released and can be re-used by the producer. When
+ * the producer and all consumers are done with the buffer, it gets destroyed.
+ */
 struct wlr_buffer {
 	const struct wlr_buffer_impl *impl;
 
-	size_t n_refs;
+	int width, height;
+
+	bool dropped;
+	size_t n_locks;
+
+	struct {
+		struct wl_signal destroy;
+		struct wl_signal release;
+	} events;
 };
 
+/**
+ * Initialize a buffer. This function should be called by producers. The
+ * initialized buffer is referenced: once the producer is done with the buffer
+ * they should call wlr_buffer_drop.
+ */
 void wlr_buffer_init(struct wlr_buffer *buffer,
-	const struct wlr_buffer_impl *impl);
+	const struct wlr_buffer_impl *impl, int width, int height);
 /**
- * Reference the buffer.
+ * Unreference the buffer. This function should be called by producers when
+ * they are done with the buffer.
  */
-struct wlr_buffer *wlr_buffer_ref(struct wlr_buffer *buffer);
+void wlr_buffer_drop(struct wlr_buffer *buffer);
```

Comment:
Release?

Reply:
Conflicts with the concept of "releasing" a buffer when all consumers are done with it (wording used in the Wayland protocol: `wl_buffer.release`).

Reply:
Eh. Don't have any better ideas.

Reply:
But maybe we should s/lock/acquire/ and s/unlock/release/?

Reply:
Also, chose to use "drop" instead of keeping "unref" so that old `wlr_buffer_unref` users don't forget to update their code to use `wlr_buffer_unlock`.

Reply:
Hm, okay.